File: backend/src/services/circuitBreaker.ts
/**
 * Circuit Breaker Service
 * Provides fault tolerance and prevents cascading failures
 */

import { EventEmitter } from 'events';
import { logger } from './loggerService';

export enum CircuitState {
  CLOSED = 'closed',     // Normal operation
  OPEN = 'open',         // Circuit is open, requests fail fast
  HALF_OPEN = 'half_open' // Testing if service has recovered
}

export interface CircuitBreakerConfig {
  failureThreshold: number;      // Number of failures before opening
  recoveryTimeout: number;       // Time to wait before trying half-open
  successThreshold: number;      // Successes needed to close from half-open
  timeout: number;              // Request timeout
  monitoringPeriod: number;     // Time window for failure counting
}

export interface CircuitBreakerStats {
  state: CircuitState;
  failureCount: number;
  successCount: number;
  lastFailureTime: Date | null;
  lastSuccessTime: Date | null;
  totalRequests: number;
  totalFailures: number;
  totalSuccesses: number;
}

export class CircuitBreaker extends EventEmitter {
  private state: CircuitState = CircuitState.CLOSED;
  private failureCount: number = 0;
  private successCount: number = 0;
  private lastFailureTime: Date | null = null;
  private lastSuccessTime: Date | null = null;
  private totalRequests: number = 0;
  private totalFailures: number = 0;
  private totalSuccesses: number = 0;
  private nextAttempt: Date = new Date();
  private config: CircuitBreakerConfig;

  constructor(
    private serviceName: string,
    config: Partial<CircuitBreakerConfig> = {}
  ) {
    super();
    
    this.config = {
      failureThreshold: config.failureThreshold || 5,
      recoveryTimeout: config.recoveryTimeout || 60000, // 1 minute
      successThreshold: config.successThreshold || 3,
      timeout: config.timeout || 30000, // 30 seconds
      monitoringPeriod: config.monitoringPeriod || 300000 // 5 minutes
    };

    logger.info(`Circuit breaker initialized for service: ${serviceName}`, {
      service: 'circuit-breaker',
      metadata: { serviceName, config: this.config }
    });
  }

  /**
   * Execute a function with circuit breaker protection
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    this.totalRequests++;

    if (this.state === CircuitState.OPEN) {
      if (Date.now() < this.nextAttempt.getTime()) {
        const error = new Error(`Circuit breaker is OPEN for service: ${this.serviceName}`);
        this.emit('requestRejected', { serviceName: this.serviceName, error });
        throw error;
      } else {
        // Try to transition to half-open
        this.state = CircuitState.HALF_OPEN;
        this.successCount = 0;
        logger.info(`Circuit breaker transitioning to HALF_OPEN: ${this.serviceName}`, {
          service: 'circuit-breaker',
          metadata: { serviceName: this.serviceName }
        });
        this.emit('stateChanged', { serviceName: this.serviceName, state: this.state });
      }
    }

    try {
      const result = await this.executeWithTimeout(fn);
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure(error as Error);
      throw error;
    }
  }

  /**
   * Execute function with timeout
   */
  private async executeWithTimeout<T>(fn: () => Promise<T>): Promise<T> {
    return new Promise((resolve, reject) => {
      const timeoutId = setTimeout(() => {
        reject(new Error(`Request timeout after ${this.config.timeout}ms`));
      }, this.config.timeout);

      fn()
        .then(result => {
          clearTimeout(timeoutId);
          resolve(result);
        })
        .catch(error => {
          clearTimeout(timeoutId);
          reject(error);
        });
    });
  }

  /**
   * Handle successful request
   */
  private onSuccess(): void {
    this.totalSuccesses++;
    this.lastSuccessTime = new Date();
    this.failureCount = 0; // Reset failure count on success

    if (this.state === CircuitState.HALF_OPEN) {
      this.successCount++;
      if (this.successCount >= this.config.successThreshold) {
        this.state = CircuitState.CLOSED;
        logger.info(`Circuit breaker CLOSED for service: ${this.serviceName}`, {
          service: 'circuit-breaker',
          metadata: { serviceName: this.serviceName }
        });
        this.emit('stateChanged', { serviceName: this.serviceName, state: this.state });
      }
    }

    this.emit('requestSucceeded', { serviceName: this.serviceName });
  }

  /**
   * Handle failed request
   */
  private onFailure(error: Error): void {
    this.totalFailures++;
    this.lastFailureTime = new Date();
    this.failureCount++;

    logger.warn(`Circuit breaker recorded failure for service: ${this.serviceName}`, {
      service: 'circuit-breaker',
      metadata: { 
        serviceName: this.serviceName, 
        failureCount: this.failureCount,
        threshold: this.config.failureThreshold
      },
      error: {
        message: error.message,
        stack: error.stack
      }
    });

    if (this.state === CircuitState.HALF_OPEN) {
      // Go back to open state
      this.state = CircuitState.OPEN;
      this.nextAttempt = new Date(Date.now() + this.config.recoveryTimeout);
      logger.warn(`Circuit breaker OPENED (from half-open) for service: ${this.serviceName}`, {
        service: 'circuit-breaker',
        metadata: { serviceName: this.serviceName }
      });
      this.emit('stateChanged', { serviceName: this.serviceName, state: this.state });
    } else if (this.state === CircuitState.CLOSED && this.failureCount >= this.config.failureThreshold) {
      // Open the circuit
      this.state = CircuitState.OPEN;
      this.nextAttempt = new Date(Date.now() + this.config.recoveryTimeout);
      logger.error(`Circuit breaker OPENED for service: ${this.serviceName}`, {
        service: 'circuit-breaker',
        metadata: { 
          serviceName: this.serviceName,
          failureCount: this.failureCount,
          threshold: this.config.failureThreshold
        }
      });
      this.emit('stateChanged', { serviceName: this.serviceName, state: this.state });
    }

    this.emit('requestFailed', { serviceName: this.serviceName, error });
  }

  /**
   * Get current circuit breaker statistics
   */
  getStats(): CircuitBreakerStats {
    return {
      state: this.state,
      failureCount: this.failureCount,
      successCount: this.successCount,
      lastFailureTime: this.lastFailureTime,
      lastSuccessTime: this.lastSuccessTime,
      totalRequests: this.totalRequests,
      totalFailures: this.totalFailures,
      totalSuccesses: this.totalSuccesses
    };
  }

  /**
   * Get current state
   */
  getState(): CircuitState {
    return this.state;
  }

  /**
   * Check if circuit breaker allows requests
   */
  canExecute(): boolean {
    if (this.state === CircuitState.CLOSED || this.state === CircuitState.HALF_OPEN) {
      return true;
    }
    
    if (this.state === CircuitState.OPEN) {
      return Date.now() >= this.nextAttempt.getTime();
    }
    
    return false;
  }

  /**
   * Manually reset the circuit breaker
   */
  reset(): void {
    this.state = CircuitState.CLOSED;
    this.failureCount = 0;
    this.successCount = 0;
    this.nextAttempt = new Date();
    
    logger.info(`Circuit breaker manually reset for service: ${this.serviceName}`, {
      service: 'circuit-breaker',
      metadata: { serviceName: this.serviceName }
    });
    
    this.emit('stateChanged', { serviceName: this.serviceName, state: this.state });
  }

  /**
   * Update configuration
   */
  updateConfig(newConfig: Partial<CircuitBreakerConfig>): void {
    this.config = { ...this.config, ...newConfig };
    
    logger.info(`Circuit breaker configuration updated for service: ${this.serviceName}`, {
      service: 'circuit-breaker',
      metadata: { serviceName: this.serviceName, config: this.config }
    });
  }
}

/**
 * Circuit Breaker Manager
 * Manages multiple circuit breakers for different services
 */
export class CircuitBreakerManager {
  private static instance: CircuitBreakerManager;
  private circuitBreakers: Map<string, CircuitBreaker> = new Map();

  private constructor() {}

  static getInstance(): CircuitBreakerManager {
    if (!CircuitBreakerManager.instance) {
      CircuitBreakerManager.instance = new CircuitBreakerManager();
    }
    return CircuitBreakerManager.instance;
  }

  /**
   * Get or create a circuit breaker for a service
   */
  getCircuitBreaker(serviceName: string, config?: Partial<CircuitBreakerConfig>): CircuitBreaker {
    if (!this.circuitBreakers.has(serviceName)) {
      const circuitBreaker = new CircuitBreaker(serviceName, config);
      this.circuitBreakers.set(serviceName, circuitBreaker);
    }
    return this.circuitBreakers.get(serviceName)!;
  }

  /**
   * Execute a function with circuit breaker protection
   */
  async execute<T>(serviceName: string, fn: () => Promise<T>, config?: Partial<CircuitBreakerConfig>): Promise<T> {
    const circuitBreaker = this.getCircuitBreaker(serviceName, config);
    return circuitBreaker.execute(fn);
  }

  /**
   * Get all circuit breaker statistics
   */
  getAllStats(): Record<string, CircuitBreakerStats> {
    const stats: Record<string, CircuitBreakerStats> = {};
    for (const [serviceName, circuitBreaker] of this.circuitBreakers.entries()) {
      stats[serviceName] = circuitBreaker.getStats();
    }
    return stats;
  }

  /**
   * Reset all circuit breakers
   */
  resetAll(): void {
    for (const circuitBreaker of this.circuitBreakers.values()) {
      circuitBreaker.reset();
    }
    
    logger.info('All circuit breakers reset', {
      service: 'circuit-breaker'
    });
  }

  /**
   * Remove a circuit breaker
   */
  removeCircuitBreaker(serviceName: string): void {
    this.circuitBreakers.delete(serviceName);
    
    logger.info(`Circuit breaker removed for service: ${serviceName}`, {
      service: 'circuit-breaker',
      metadata: { serviceName }
    });
  }
}

export const circuitBreakerManager = CircuitBreakerManager.getInstance();

File: backend-fastapi/app/services/circuit_breaker.py
"""
Circuit Breaker Service
Fault tolerance and resilience patterns for external service calls
"""

import asyncio
import time
from enum import Enum
from typing import Any, Callable, Dict, Optional, Union
from dataclasses import dataclass, field

from circuitbreaker import CircuitBreaker as BaseCircuitBreaker

from app.core.config import settings
from app.core.logging import logger
from app.core.exceptions import CircuitBreakerOpenError, ExternalServiceError


class CircuitBreakerState(Enum):
    """Circuit breaker states"""
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


@dataclass
class CircuitBreakerStats:
    """Circuit breaker statistics"""
    name: str
    state: CircuitBreakerState
    failure_count: int = 0
    success_count: int = 0
    total_calls: int = 0
    last_failure_time: Optional[float] = None
    last_success_time: Optional[float] = None
    recovery_timeout: int = 60
    failure_threshold: int = 5
    success_threshold: int = 3
    call_timeout: int = 30
    created_at: float = field(default_factory=time.time)


class CircuitBreakerManager:
    """Manages multiple circuit breakers for different services"""
    
    def __init__(self):
        self.circuit_breakers: Dict[str, BaseCircuitBreaker] = {}
        self.stats: Dict[str, CircuitBreakerStats] = {}
        self.default_config = settings.circuit_breaker_config
    
    def get_or_create_circuit_breaker(
        self,
        name: str,
        failure_threshold: Optional[int] = None,
        recovery_timeout: Optional[int] = None,
        expected_exception: Optional[tuple] = None
    ) -> BaseCircuitBreaker:
        """Get existing circuit breaker or create new one"""
        
        if name not in self.circuit_breakers:
            # Create new circuit breaker
            failure_threshold = failure_threshold or self.default_config["failure_threshold"]
            recovery_timeout = recovery_timeout or self.default_config["recovery_timeout"]
            expected_exception = expected_exception or self.default_config["expected_exception"]
            
            circuit_breaker = BaseCircuitBreaker(
                failure_threshold=failure_threshold,
                recovery_timeout=recovery_timeout,
                expected_exception=expected_exception
            )
            
            # Add event listeners
            circuit_breaker.add_listener(self._on_circuit_breaker_open(name))
            circuit_breaker.add_listener(self._on_circuit_breaker_close(name))
            circuit_breaker.add_listener(self._on_circuit_breaker_half_open(name))
            
            self.circuit_breakers[name] = circuit_breaker
            
            # Initialize stats
            self.stats[name] = CircuitBreakerStats(
                name=name,
                state=CircuitBreakerState.CLOSED,
                failure_threshold=failure_threshold,
                recovery_timeout=recovery_timeout
            )
            
            logger.info(
                f"Created circuit breaker: {name}",
                service="circuit_breaker",
                metadata={
                    "name": name,
                    "failure_threshold": failure_threshold,
                    "recovery_timeout": recovery_timeout
                }
            )
        
        return self.circuit_breakers[name]
    
    def _on_circuit_breaker_open(self, name: str):
        """Callback when circuit breaker opens"""
        def callback():
            if name in self.stats:
                self.stats[name].state = CircuitBreakerState.OPEN
                self.stats[name].last_failure_time = time.time()
            
            logger.warning(
                f"Circuit breaker opened: {name}",
                service="circuit_breaker",
                metadata={
                    "name": name,
                    "failure_count": self.stats[name].failure_count if name in self.stats else 0
                }
            )
        
        return callback
    
    def _on_circuit_breaker_close(self, name: str):
        """Callback when circuit breaker closes"""
        def callback():
            if name in self.stats:
                self.stats[name].state = CircuitBreakerState.CLOSED
                self.stats[name].last_success_time = time.time()
            
            logger.info(
                f"Circuit breaker closed: {name}",
                service="circuit_breaker",
                metadata={
                    "name": name,
                    "success_count": self.stats[name].success_count if name in self.stats else 0
                }
            )
        
        return callback
    
    def _on_circuit_breaker_half_open(self, name: str):
        """Callback when circuit breaker goes half-open"""
        def callback():
            if name in self.stats:
                self.stats[name].state = CircuitBreakerState.HALF_OPEN
            
            logger.info(
                f"Circuit breaker half-open: {name}",
                service="circuit_breaker",
                metadata={"name": name}
            )
        
        return callback
    
    async def call_with_circuit_breaker(
        self,
        name: str,
        func: Callable,
        *args,
        failure_threshold: Optional[int] = None,
        recovery_timeout: Optional[int] = None,
        **kwargs
    ) -> Any:
        """Execute function with circuit breaker protection"""
        
        circuit_breaker = self.get_or_create_circuit_breaker(
            name, failure_threshold, recovery_timeout
        )
        
        # Update stats
        if name in self.stats:
            self.stats[name].total_calls += 1
        
        try:
            # Check if circuit breaker is open
            if circuit_breaker.current_state == "open":
                raise CircuitBreakerOpenError(name)
            
            # Execute function with circuit breaker
            if asyncio.iscoroutinefunction(func):
                result = await circuit_breaker(func)(*args, **kwargs)
            else:
                result = circuit_breaker(func)(*args, **kwargs)
            
            # Update success stats
            if name in self.stats:
                self.stats[name].success_count += 1
                self.stats[name].last_success_time = time.time()
            
            return result
            
        except CircuitBreakerOpenError:
            raise
        except Exception as e:
            # Update failure stats
            if name in self.stats:
                self.stats[name].failure_count += 1
                self.stats[name].last_failure_time = time.time()
            
            logger.error(
                f"Circuit breaker call failed: {name}",
                service="circuit_breaker",
                error={"message": str(e), "type": type(e).__name__},
                metadata={"name": name, "function": func.__name__ if hasattr(func, '__name__') else str(func)}
            )
            
            raise ExternalServiceError(name, str(e))
    
    def get_circuit_breaker_state(self, name: str) -> Optional[CircuitBreakerState]:
        """Get current state of circuit breaker"""
        if name in self.stats:
            return self.stats[name].state
        return None
    
    def get_circuit_breaker_stats(self, name: str) -> Optional[CircuitBreakerStats]:
        """Get statistics for specific circuit breaker"""
        return self.stats.get(name)
    
    def get_all_stats(self) -> Dict[str, Dict[str, Any]]:
        """Get statistics for all circuit breakers"""
        result = {}
        
        for name, stats in self.stats.items():
            circuit_breaker = self.circuit_breakers.get(name)
            
            result[name] = {
                "name": stats.name,
                "state": stats.state.value,
                "failure_count": stats.failure_count,
                "success_count": stats.success_count,
                "total_calls": stats.total_calls,
                "success_rate": (stats.success_count / stats.total_calls * 100) if stats.total_calls > 0 else 0,
                "failure_rate": (stats.failure_count / stats.total_calls * 100) if stats.total_calls > 0 else 0,
                "last_failure_time": stats.last_failure_time,
                "last_success_time": stats.last_success_time,
                "failure_threshold": stats.failure_threshold,
                "recovery_timeout": stats.recovery_timeout,
                "uptime_seconds": time.time() - stats.created_at,
                "current_state": circuit_breaker.current_state if circuit_breaker else "unknown"
            }
        
        return result
    
    def reset_circuit_breaker(self, name: str) -> bool:
        """Reset circuit breaker to closed state"""
        if name in self.circuit_breakers:
            circuit_breaker = self.circuit_breakers[name]
            
            # Reset the circuit breaker
            circuit_breaker._failure_count = 0
            circuit_breaker._last_failure = None
            circuit_breaker._state = "closed"
            
            # Reset stats
            if name in self.stats:
                self.stats[name].state = CircuitBreakerState.CLOSED
                self.stats[name].failure_count = 0
                self.stats[name].last_failure_time = None
            
            logger.info(
                f"Circuit breaker reset: {name}",
                service="circuit_breaker",
                metadata={"name": name}
            )
            
            return True
        
        return False
    
    def remove_circuit_breaker(self, name: str) -> bool:
        """Remove circuit breaker"""
        removed = False
        
        if name in self.circuit_breakers:
            del self.circuit_breakers[name]
            removed = True
        
        if name in self.stats:
            del self.stats[name]
            removed = True
        
        if removed:
            logger.info(
                f"Circuit breaker removed: {name}",
                service="circuit_breaker",
                metadata={"name": name}
            )
        
        return removed
    
    def get_health_summary(self) -> Dict[str, Any]:
        """Get health summary of all circuit breakers"""
        total_breakers = len(self.stats)
        open_breakers = sum(1 for stats in self.stats.values() if stats.state == CircuitBreakerState.OPEN)
        half_open_breakers = sum(1 for stats in self.stats.values() if stats.state == CircuitBreakerState.HALF_OPEN)
        closed_breakers = total_breakers - open_breakers - half_open_breakers
        
        return {
            "total_circuit_breakers": total_breakers,
            "closed": closed_breakers,
            "open": open_breakers,
            "half_open": half_open_breakers,
            "health_status": "healthy" if open_breakers == 0 else "degraded" if open_breakers < total_breakers else "unhealthy",
            "open_breakers": [name for name, stats in self.stats.items() if stats.state == CircuitBreakerState.OPEN]
        }


# Decorator for circuit breaker protection
def circuit_breaker(
    name: str,
    failure_threshold: Optional[int] = None,
    recovery_timeout: Optional[int] = None
):
    """Decorator to add circuit breaker protection to functions"""
    def decorator(func):
        async def wrapper(*args, **kwargs):
            return await circuit_breaker_manager.call_with_circuit_breaker(
                name, func, *args,
                failure_threshold=failure_threshold,
                recovery_timeout=recovery_timeout,
                **kwargs
            )
        return wrapper
    return decorator


# Global circuit breaker manager instance
circuit_breaker_manager = CircuitBreakerManager()